braze/client.py

```diff
@@ -81,41 +92,80 @@ def user_delete(self, external_ids, appboy_ids):
 
-        response = {}
-
+        response = {'errors': []}
         try:
-            if request_type == self.REQUEST_POST:
-                r = self.requests.post(self.request_url, data=json.dumps(payload), headers=self.headers)
-                response = r.json()
-                response['status_code'] = r.status_code
-                if response['message'] == 'success' and 'errors' not in response:
+            r = self._post_request_with_retries(payload)
+            response.update(r.json())
+            response['status_code'] = r.status_code
+
+            message = response['message']
+            if message == 'success' or message == 'queued':
+                if not response['errors']:
                     response['success'] = True
+                else:
+                    # Non-Fatal errors
+                    pass
 
-        except RequestException as e:
-            # handle all requests HTTP exceptions
-            response = {'client_error': str(e)}
-        except Exception as e:
-            # handle all exceptions which can be on API side
-            response = {'client_error': (str(e) + '. Response: ' + r.text)}
+        except (
+                RequestException,
+                BrazeRateLimitError,
+                BrazeInternalServerError,
+        ) as e:
+            response['errors'].append(str(e))
 
         if 'success' not in response:
             response['success'] = False
-        if 'errors' not in response:
-            response['errors'] = ''
+
         if 'status_code' not in response:
             response['status_code'] = 0
+
         if 'message' not in response:
             response['message'] = ''
-        if 'client_error' not in response:
-            response['client_error'] = ''
 
         return response
+
+    @retry(
+        reraise=True,
+        wait=wait_random_exponential(multiplier=1, max=MAX_WAIT_SECONDS),
+        stop=stop_after_attempt(MAX_RETRIES),
+        retry=(
+            retry_if_exception_type(BrazeInternalServerError)
+            | retry_if_exception_type(RequestException)
+        ),
+    )
+    def _post_request_with_retries(self, payload, retry_attempt=0):
+        """
+        :param dict payload:
+        :param int retry_attempt: current retry attempt number
+        :rtype: dict
+        """
+        headers = {
+            'Content-Type': 'application/json',
+        }
+        r = requests.post(
+            self.request_url,
+            data=json.dumps(payload),
+            headers=headers,
+            timeout=2,
+        )
+        if retry_attempt >= self.MAX_RETRIES:
+            raise BrazeRateLimitError('BrazeRateLimitError')
+
+        # https://www.braze.com/docs/developer_guide/rest_api/messaging/#fatal-errors
+        if r.status_code == 429:
+            reset_epoch_seconds = float(r.headers.get('X-RateLimit-Reset'))
+            sec_to_reset = reset_epoch_seconds - float(time.time())
+            if sec_to_reset < self.MAX_WAIT_SECONDS:
```

Comment:
We should diff against time budget remaining i.e. count the time spent on the first request as part of the budget already spent.

Reply:
Any issues with wait time per request? That's the pattern tenacity uses as well .  If total maximum wait time is a concern then it should be noted that we do have an upper cap on total maximum wait time which is MAX_WAIT_SECONDS * (MAX_RETRIES + 1). We can tweak MAX_WAIT_SECONDS or MAX_RETRIES to reduce total maximum wait time for the worst case scenario.

Reply:
MAX_WAIT_SECONDS updated to `1.25` and max retries at 3 so now,  total max wait time (in worst case) = MAX_WAIT_SECONDS * (MAX_RETRIES + 1) = 5 seconds.